tests/unit/tuf_repository_service_worker/models/targets/test_crud.py

```diff
@@ -0,0 +1,281 @@
+# SPDX-FileCopyrightText: 2022 VMware Inc
+#
+# SPDX-License-Identifier: MIT
+import datetime
+
+import pretend
+
+from repository_service_tuf_worker.models.targets import crud
+
+
+class TestTargetsCrud:
+    def test_create(self):
+        crud.models.RSTUFTargets = pretend.call_recorder(
+            lambda *a, **kw: "fake_db_target"
+        )
+        mocked_db = pretend.stub(
+            add=pretend.call_recorder(lambda *a: None),
+            commit=pretend.call_recorder(lambda: None),
+            refresh=pretend.call_recorder(lambda *a: None),
+        )
+        test_target = crud.schemas.TargetsCreate(
+            path="file1.tar.gz",
+            info={"info": {"k": "v"}},
+            rolename="bins-0",
+            published=False,
+            action=crud.schemas.TargetAction.ADD,
+        )
+
+        test_result = crud.create(mocked_db, test_target)
+        assert test_result == "fake_db_target"
+        assert mocked_db.add.calls == [pretend.call("fake_db_target")]
+        assert mocked_db.commit.calls == [pretend.call()]
+        assert mocked_db.refresh.calls == [pretend.call("fake_db_target")]
+
+    def test_read_unpublished_rolenames(self):
+
+        crud.models.RSTUFTargets = pretend.stub(
+            published=False, rolename="all-bins"
+        )
+
+        mocked_distinct = pretend.stub(
+            all=pretend.call_recorder(
+                lambda: [(False, "bin-e"), (False, "bin-3")]
+            )
+        )
+        mocked_filter_response = pretend.stub(
+            distinct=pretend.call_recorder(lambda: mocked_distinct)
+        )
+        mocked_db_filter = pretend.stub(
+            filter=pretend.call_recorder(lambda *a: mocked_filter_response)
+        )
+        mocked_db = pretend.stub(
+            query=pretend.call_recorder(lambda *a: mocked_db_filter)
+        )
+
+        test_result = crud.read_unpublished_rolenames(mocked_db)
+
+        assert test_result == [(False, "bin-e"), (False, "bin-3")]
+        assert mocked_db.query.calls == [pretend.call(False, "all-bins")]
+        assert mocked_db_filter.filter.calls == [pretend.call(True)]
```

Comment:
I am sorry, but I didn't see your response to this question.

Reply:
OK, I left it to discuss in a call, but writing here.

We are testing `test_read_unpublished_rolenames`.
It uses the SQLAchemy and we use the db session created by it that has the built-ins `query`, `filter` etc.

The [pretend.](https://github.com/alex/pretend)`call_recorder` recorders the call parameters used by the function, not the return. That way, you can track each unit of your code and see if it is calling corrected, even if it is mocked.

The `mocked_db_filter.filter.calls` will have a "pretend call" with `True` because we use pretend for mocking it. We are not testing what SQLAlchemy is using here.
We could even get ride-off `mock_db_dilter.calls` here, to be honest. The important is the query call.